server/src/main/java/org/apache/druid/server/http/MetadataResource.java

```diff
@@ -171,26 +175,67 @@ public Response getDatabaseSegments(
+      //transform DataSegment to SegmentWithOvershadowedStatus objects
+      final Stream<SegmentWithOvershadowedStatus> segmentsWithOvershadowedStatus = metadataSegments
+          .map(
+              segment -> new SegmentWithOvershadowedStatus(
+                  segment,
+                  overshadowedSegments.contains(segment.getId())
+              )).collect(Collectors.toList()).stream();
 
-    final Iterable<DataSegment> authorizedSegments =
-        AuthorizationUtils.filterAuthorizedResources(req, metadataSegments::iterator, raGenerator, authorizerMapper);
+      final Function<SegmentWithOvershadowedStatus, Iterable<ResourceAction>> raGenerator = segment -> Collections.singletonList(
+          AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR.apply(segment.getDataSegment().getDataSource()));
 
-    final StreamingOutput stream = outputStream -> {
-      final JsonFactory jsonFactory = jsonMapper.getFactory();
-      try (final JsonGenerator jsonGenerator = jsonFactory.createGenerator(outputStream)) {
-        jsonGenerator.writeStartArray();
-        for (DataSegment ds : authorizedSegments) {
-          jsonGenerator.writeObject(ds);
-          jsonGenerator.flush();
-        }
-        jsonGenerator.writeEndArray();
-      }
-    };
+      final Iterable<SegmentWithOvershadowedStatus> authorizedSegments =
+          AuthorizationUtils.filterAuthorizedResources(
+              req,
+              segmentsWithOvershadowedStatus::iterator,
+              raGenerator,
+              authorizerMapper
+          );
+      Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+      return builder.entity(authorizedSegments).build();
+    } else {
 
-    Response.ResponseBuilder builder = Response.status(Response.Status.OK);
-    return builder.entity(stream).build();
+      final Function<DataSegment, Iterable<ResourceAction>> raGenerator = segment -> Collections.singletonList(
+          AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR.apply(segment.getDataSource()));
+
+      final Iterable<DataSegment> authorizedSegments =
+          AuthorizationUtils.filterAuthorizedResources(req, metadataSegments::iterator, raGenerator, authorizerMapper);
+
+      Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+      return builder.entity(authorizedSegments).build();
+    }
+  }
+
+  /**
+   * find fully overshadowed segments
+   *
+   * @param druidDataSources
+   *
+   * @return set of overshadowed segments
+   */
+  private Set<SegmentId> findOvershadowedSegments(Collection<ImmutableDruidDataSource> druidDataSources)
+  {
+    final Stream<DataSegment> segmentStream = druidDataSources
+        .stream()
+        .flatMap(t -> t.getSegments().stream());
+    final Set<DataSegment> usedSegments = segmentStream.collect(Collectors.toSet());
+    final Map<String, VersionedIntervalTimeline<String, DataSegment>> timelines = new HashMap<>();
```

Comment:
This timelines building procedure appears in several other places in code, can it be extracted as a method?

Reply:
yeah, extracted it to `VersionedIntervalTimeline`, added static method `buildTimelines()`